bpf/rtt_tracker.h

```diff
@@ -0,0 +1,69 @@
+#ifndef __RTT_TRACKER_H__
+#define __RTT_TRACKER_H__
+
+#include "utils.h"
+#include "maps_definition.h"
+
+static __always_inline void fill_flow_seq_id(flow_seq_id *seq_id, pkt_info *pkt, u32 seq, u8 reversed) {
+    flow_id *id = pkt->id;
+    if (reversed) {
+        __builtin_memcpy(seq_id->src_ip, id->dst_ip, IP_MAX_LEN);
+        __builtin_memcpy(seq_id->dst_ip, id->src_ip, IP_MAX_LEN);
+        seq_id->src_port = id->dst_port;
+        seq_id->dst_port = id->src_port;
+    } else {
+        __builtin_memcpy(seq_id->src_ip, id->src_ip, IP_MAX_LEN);
+        __builtin_memcpy(seq_id->dst_ip, id->dst_ip, IP_MAX_LEN);
+        seq_id->src_port = id->src_port;
+        seq_id->dst_port = id->dst_port;
+    }
+    seq_id->seq_id = seq;
+}
+
+static __always_inline void calculate_flow_rtt(pkt_info *pkt, u8 direction, void *data_end) {
+    flow_seq_id seq_id;
+    __builtin_memset(&seq_id, 0, sizeof(flow_seq_id));
+
+    switch (pkt->id->transport_protocol)
+    {
+    case IPPROTO_TCP: {
+            struct tcphdr *tcp = (struct tcphdr *) pkt->l4_hdr;
+            if ( !tcp || ((void *)tcp + sizeof(*tcp) > data_end) ) {
+                break;
+            }
+            if ((direction == EGRESS) && IS_SYN_PACKET(pkt)) {
+                // Record the outgoing syn sequence number
+                u32 seq = bpf_ntohl(tcp->seq);
+                fill_flow_seq_id(&seq_id, pkt, seq, 0);
+
+                long ret = bpf_map_update_elem(&flow_sequences, &seq_id, &pkt->current_ts, BPF_ANY);
+                if (trace_messages && ret != 0) {
+                    bpf_printk("err saving flow sequence record %d", ret);
+                }
+                break;
+            }
+            if ((direction == INGRESS) && IS_ACK_PACKET(pkt)) {
+                // Stored sequence should be ack_seq - 1
+                u32 seq = bpf_ntohl(tcp->ack_seq) - 1;
+                // check reversed flow
+                fill_flow_seq_id(&seq_id, pkt, seq, 1); 
```

Comment:
what happens when there are retransimits ?

Reply:
In case of retransmissions we have two scenarios, if SYN is retransmitted the last one will be taken into account which is the correct behavior.
If ACK is retransmitted the receiver will only see the last ACK which it received and for the receiver the calculated RTT will be bigger than the line value but this behavior will be correct because the receiver doesn't know about ACK retransmission by the sender. 
This will be a momentary spike in the reading which will hopefully be correct when we enable continous rtt calculations for a flow.

This spike in the reading can be later correlated with tcp drops or something else as well.

Reply:
ok for now pls doc this behavior so we have reference in case we need to revisit

Reply:
Added.